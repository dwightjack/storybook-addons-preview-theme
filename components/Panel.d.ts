import * as React from 'react';
export interface PanelProp {
    api: any;
    active: boolean;
    channel: any;
}
export declare function Panel({ api, active, channel, }: {
    api: any;
    active: any;
    channel: any;
}): React.FunctionComponentElement<PanelProp>;
