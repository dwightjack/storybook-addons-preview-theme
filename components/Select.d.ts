import * as React from 'react';
import { Theme } from '../types';
export interface SelectProps {
    onChange: () => void;
    options: Theme[];
    value?: string;
}
export declare function Select({ onChange, options, value, }: {
    onChange: any;
    options: any;
    value?: string;
}): React.FunctionComponentElement<SelectProps>;
