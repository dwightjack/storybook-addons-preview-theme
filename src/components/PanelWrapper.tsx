import { styled } from '@storybook/theming';

export const PanelWrapper = styled.div`
  height: 100%;
  overflow: auto;
  width: 100%;
`;
