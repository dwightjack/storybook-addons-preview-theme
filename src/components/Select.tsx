import * as React from 'react';
import { Form } from '@storybook/components';
import { Theme } from '../types';

export interface SelectProps {
  onChange: () => void;
  options: Theme[];
  value?: string;
}

export function Select({
  onChange,
  options,
  value = '',
}): React.FunctionComponentElement<SelectProps> {
  return (
    <Form.Select
      onChange={onChange}
      value={value}
      name="preview-theme"
      label="Preview Theme"
      size="flex"
    >
      {[
        { name: 'empty', className: 'none', label: 'select...' },
        ...options,
      ].map(({ name, label, className }) => (
        <option value={className} key={name}>
          {label}
        </option>
      ))}
    </Form.Select>
  );
}
