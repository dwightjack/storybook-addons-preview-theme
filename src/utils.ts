import { Theme, ThemeProp, ThemeConfig } from './types';

export function isEmpty(arr: any[]): boolean {
  if (!Array.isArray(arr)) {
    return true;
  }
  return arr.length <= 0;
}

export const getThemes = ({
  globalThemes = [],
  themes,
}: ThemeConfig): Theme[] | null => {
  if (isEmpty(globalThemes) || !themes) {
    return null;
  }
  if (themes === true && !isEmpty(globalThemes)) {
    return globalThemes;
  }
  return (themes as ThemeProp[])
    .map((theme) => {
      if (typeof theme === 'string') {
        return globalThemes.find(({ name }) => name === theme);
      }
      return theme;
    })
    .filter((x) => x);
};
