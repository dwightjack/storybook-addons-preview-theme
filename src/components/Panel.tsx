import * as React from 'react';
import { useState, useEffect } from 'react';
import { Form, Placeholder } from '@storybook/components';
import { STORY_CHANGED } from '@storybook/core-events';
import { Select } from './Select';
import { PanelWrapper } from './PanelWrapper';
import { getThemes } from '../utils';

export interface PanelProp {
  api: any;
  active: boolean;
  channel: any;
}

export function Panel({
  api,
  active,
  channel,
}): React.FunctionComponentElement<PanelProp> {
  const [selected, setSelected] = useState('none');
  const data = api.getCurrentStoryData();

  useEffect(() => {
    const listener = (): void => {
      setSelected('none');
    };

    api.on(STORY_CHANGED, listener);

    return () => {
      api.off(STORY_CHANGED, listener);
    };
  }, []);

  useEffect(() => {
    channel.emit('preview-theme:change', selected);
  }, [selected]);

  if (!active || !data) {
    return null;
  }

  const themes = getThemes(data.parameters || {});

  if (!Array.isArray(themes) || themes.length === 0) {
    return (
      <Placeholder>Theme swatches not available for this story</Placeholder>
    );
  }

  return (
    <PanelWrapper>
      <Form>
        <Form.Field>
          <Select
            value={selected}
            onChange={({ target }) => setSelected(target.value)}
            options={themes}
          />
        </Form.Field>
      </Form>
    </PanelWrapper>
  );
}
