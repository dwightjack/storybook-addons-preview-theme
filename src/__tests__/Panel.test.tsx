import * as React from 'react';
import { render, fireEvent } from 'react-testing-library';
import * as components from '@storybook/components';
import { STORY_CHANGED } from '@storybook/core-events';
import { Panel } from '../components/Panel';
import * as utils from '../utils';

jest.mock('../utils', () => {
  return {
    getThemes: jest.fn(),
  };
});

jest.mock('@storybook/components', () => {
  const Placeholder = jest.fn(() => <div />);
  const Form = ({ children }) => <form>{children}</form>;
  Form.Field = ({ children }) => <div>{children}</div>;
  Form.Select = ({ children, ...props }) => {
    return <input className="select" {...props} />;
  };
  return { Form, Placeholder };
});

describe('<Panel />', () => {
  let api;
  let channel;
  beforeEach(() => {
    channel = { emit: jest.fn() };
    api = {
      getCurrentStoryData: jest.fn(() => ({})),
      on: jest.fn(),
      off: jest.fn(),
    };
  });
  test('returns nothing if the panel is not active', () => {
    const { container } = render(
      <Panel api={api} channel={channel} active={false} />,
    );
    expect(container.childElementCount).toBe(0);
  });

  test('returns nothing if there are no story data', () => {
    api.getCurrentStoryData.mockImplementation(() => null);
    const { container } = render(
      <Panel api={api} channel={channel} active={true} />,
    );
    expect(container.childElementCount).toBe(0);
  });

  test('default snapshot', () => {
    render(<Panel api={api} channel={channel} active={true} />);
    expect(components.Placeholder).toHaveBeenCalled();

    components.Placeholder.mockClear();

    (utils.getThemes as any).mockImplementation(() => []);
    render(<Panel api={api} channel={channel} active={true} />);
    expect(components.Placeholder).toHaveBeenCalled();
  });

  test('default snapshot', () => {
    (utils.getThemes as any).mockImplementation(() => [{ name: 'dark' }]);
    const { container } = render(
      <Panel api={api} channel={channel} active={true} />,
    );
    expect(container.firstElementChild).toMatchSnapshot();
  });

  describe('Events and effects', () => {
    let result;
    let container;
    beforeEach(() => {
      (utils.getThemes as any).mockImplementation(() => [
        { name: 'dark', label: 'Dark', className: '-dark' },
      ]);
      result = render(<Panel api={api} channel={channel} active={true} />);
      ({ container } = result);
      channel.emit.mockClear();
    });

    test('change event handler', () => {
      fireEvent.change(container.querySelector('input.select'), {
        target: { value: 'dark' },
      });
      expect(
        container.querySelector('input.select').getAttribute('value'),
      ).toBe('dark');
    });

    test('emits "preview-theme:change"', () => {
      fireEvent.change(container.querySelector('input.select'), {
        target: { value: 'dark' },
      });
      expect(channel.emit).toHaveBeenCalledWith('preview-theme:change', 'dark');
    });

    test('attaches a listener to "STORY_CHANGED"', () => {
      expect(api.on).toHaveBeenCalledWith(STORY_CHANGED, expect.any(Function));
    });

    test('calls off with "STORY_CHANGED" on unmount ', () => {
      result.unmount();
      expect(api.off).toHaveBeenCalledWith(STORY_CHANGED, expect.any(Function));
    });
  });
});
