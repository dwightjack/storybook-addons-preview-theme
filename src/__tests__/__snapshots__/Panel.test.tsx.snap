// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Panel /> default snapshot 1`] = `
<div
  class="css-13bbivi"
>
  <form>
    <div>
      <input
        class="select"
        label="Preview Theme"
        name="preview-theme"
        value="none"
      />
    </div>
  </form>
</div>
`;
